Pet/Guardian AI hook re-organizing

Treeston opened the pull request:
Pet/Guardian AI hook re-organizing:
* Adjust OwnerAttacked/OwnerAttackedBy hooks on CreatureAI to fire for all owned units, not just player pets. This should allow guardians to more reliably recognize valid targets.
* Kill off the AttackedBy hook. While it was defined in CreatureAI.h as virtual, it was only ever invoked for player pets in specific situations. This makes it classic developer bait.
  * Adjust PetAI to use DamageTaken instead of AttackedBy.
  * Adjust behavior of AttackStart on PetAI to compensate.

Treeston committed:
Pet/Guardian AI hook re-organizing:
- Adjust OwnerAttacked/OwnerAttackedBy hooks on CreatureAI to fire for all owned units, not just player pets. This should allow guardians to more reliably recognize valid targets.
- Kill off the AttackedBy hook. While it was defined in CreatureAI.h as virtual, it was only ever invoked for player pets in specific situations. This makes it classic developer bait.
  - Adjust PetAI to use DamageTaken instead of AttackedBy.
  - Adjust behavior of AttackStart on PetAI to compensate.

Shauren commented:
Shouldn't AttackedBy fire on events that cause aggro but not deal damage like taunt/debuff/dot without instant damage component?

Treeston commented:
In current core, EnterCombat is used if it's initial aggro, otherwise spell related hooks.
If we want AttackedBy to actually be a functional hook, that'd be a different PR - in its current state, it does nothing.

Shauren commented:
Well, all of the scenarios I described are valid even as non-initial-aggro

Treeston commented:
Yes, but currently `AttackedBy` won't fire in these scenarios.

Treeston committed:
Add default implementation of OwnerAttackedBy in CreatureAI. I shouldn't trust @ccrs.

ccrs commented on src/server/game/AI/CreatureAI.cpp:
```diff
@@ -132,15 +132,22 @@ void CreatureAI::MoveInLineOfSight(Unit* who)
+    if (!attacker || !me->IsAlive())
+        return;
+
+    if (!me->HasReactState(REACT_PASSIVE) && me->CanStartAttack(attacker, false))
```
````
me->CanStartAttack(attacker, false)
````
prevents combat with neutral units that the owner is attacking
````
me->CanStartAttack(attacker, true)
````
testing with force = true

ccrs replied:
actually makes sense for it to be true, faction check can be sort of ignored and distance check is useless

ccrs replied:
"sort of ignored"

Treeston committed:
@ccrs, I trusted you. You were the chosen one.

You were supposed to tell me whether it worked without force, not lead me into darkness.

Shauren commented on src/server/game/AI/CreatureAI.h:
```diff
@@ -152,15 +150,14 @@ class TC_GAME_API CreatureAI : public UnitAI
 
         // Called when owner attacks something
-        virtual void OwnerAttacked(Unit* /*target*/) { }
+        virtual void OwnerAttacked(Unit* target) { CreatureAI::OwnerAttackedBy(target); }
```
You talked about traps for scripters but here you add another one - it would be better to introduce a separate (nonvirtual) method that is called by both OwnerAttacked and OwnerAttackedBy. Someone less experienced might think that if he overrides OwnerAttackedBy in his own script but not OwnerAttacked that it will call his overriden OwnerAttackedBy (but it will be the base CreatureAI one)

Treeston committed:
Refactor as per shauren
